providers/implementations/keymgmt/ec_kmgmt.c

```diff
@@ -767,21 +777,210 @@ int ec_validate(void *keydata, int selection)
+        gctx->gen_group = NULL;
+        gctx->keytype = "EC";
+        gctx->selection = selection;
+    }
+    return gctx;
+}
+
+static void *sm2_gen_init(void *provctx, int selection)
+{
+    struct ec_gen_ctx *gctx = ec_gen_init(provctx, selection);
+
+    if (gctx != NULL) {
+        gctx->gen_group = EC_GROUP_new_by_curve_name(NID_sm2);
+        gctx->keytype = "SM2";
+    }
+    return gctx;
+}
+
+static int ec_gen_set_group(void *genctx, int nid)
+{
+    struct ec_gen_ctx *gctx = genctx;
+    EC_GROUP *group;
+
+    /* Default keytype depending on group */
+    if (nid == NID_sm2)
+        gctx->keytype = "SM2";
+    else
+        gctx->keytype = "EC";
+
+    group = EC_GROUP_new_by_curve_name(nid); /* Odd function name... */
+    if (group == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_INVALID_CURVE);
+        return 0;
+    }
+    EC_GROUP_free(gctx->gen_group);
+    gctx->gen_group = group;
+    return 1;
+}
+static int ec_gen_set_template(void *genctx, void *templ)
+{
+    struct ec_gen_ctx *gctx = genctx;
+    EC_KEY *ec = templ;
+    const EC_GROUP *ec_group;
+
+    if (gctx == NULL || ec == NULL)
+        return 0;
+    if ((ec_group = EC_KEY_get0_group(ec)) == NULL)
+        return 0;
+    return ec_gen_set_group(gctx, EC_GROUP_get_curve_name(ec_group));
+}
+
+static int ec_gen_set_params(void *genctx, const OSSL_PARAM params[])
+{
+    struct ec_gen_ctx *gctx = genctx;
+    const OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate_const(params, OSSL_PKEY_PARAM_EC_NAME))
+        != NULL) {
+        const char *curve_name = NULL;
+
+        if (!OSSL_PARAM_get_utf8_ptr(p, &curve_name))
+            if (p->data_type == OSSL_PARAM_UTF8_STRING)
+                curve_name = p->data;
+        if (curve_name == NULL)
+            return 0;
+
+        if (!ec_gen_set_group(gctx,  ec_curve_name2nid(curve_name)))
+            return 0;
+    }
+    return 1;
+}
+
+static const OSSL_PARAM *ec_gen_settable_params(void *provctx)
+{
+    static OSSL_PARAM settable[] = {
+        { OSSL_PKEY_PARAM_EC_NAME, OSSL_PARAM_UTF8_PTR, NULL, 0, 0 },
+        OSSL_PARAM_END
+    };
+
+    return settable;
+}
+
+static int ec_gen_get_params(void *genctx, OSSL_PARAM params[])
+{
+    struct ec_gen_ctx *gctx = genctx;
+    OSSL_PARAM *p;
+
+    if ((p = OSSL_PARAM_locate(params, OSSL_PKEY_PARAM_KEYTYPE)) == NULL
+        || !OSSL_PARAM_set_utf8_ptr(p, gctx->keytype))
+        return 0;
+    return 1;
+}
+
+static const OSSL_PARAM *ec_gen_gettable_params(void *provctx)
+{
+    static OSSL_PARAM gettable[] = {
+        { OSSL_PKEY_PARAM_KEYTYPE, OSSL_PARAM_UTF8_PTR, NULL, 0, 0 },
+        OSSL_PARAM_END
+    };
+
+    return gettable;
+}
+
+static int ec_gen_assign_group(EC_KEY *ec, EC_GROUP *group)
+{
+    if (group == NULL) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_NO_PARAMETERS_SET);
+        return 0;
+    }
+    return EC_KEY_set_group(ec, group) > 0;
+}
+
+/*
+ * No callback defined for EC_KEY generation, so we don't use them.
```

Comment:
Or you could just say :)

The callback arguments (osslcb & cbarg) are not used by EC_KEY generation